server/pulp/server/db/model/__init__.py

```diff
@@ -20,7 +21,23 @@
 _logger = logging.getLogger(__name__)
 
 
-class Repository(Document):
+class AutoRetryDocument(Document):
+    """
+    Base class for mongoengine documents, includes auto retry functionality,
+    if unsafe_autoretry is set to true in the server config.
+    """
+
+    def __init__(self, *args, **kwargs):
+        """
+        Initialize a document and decorate the appropriate methods with the retry_decorator.
+        """
+        super(AutoRetryDocument, self).__init__(*args, **kwargs)
+        UnsafeRetry.decorate_instance(instance=self, full_name=type(self))
+
+    meta = {'abstract': True}
```

Comment:
@bmbouter, I tried both ways, allowing inheritance or to declare an abstract document. They each work equally well, but I read up a bit, and `allow_inheritance` seems to be designed for Documents that can be instantiated but may also allow others to be subclasses of them (in the same collection). Abstract documents are simple base documents that cannot be instantiated on their own but can still be the parent class of other documents-- this seemed to be more appropriate.